include/serialport.h

```diff
@@ -47,116 +50,143 @@
 #include "hardware.h"
 #endif
 
-// Serial port interface 
+// Serial port interface
+
+#define SERIAL_MAX_FIFO_SIZE 256
+/* Note: Almost all DOS-era universal asynchronous receiver-transmitter
+ *       (UART)'s permitted up to a 16-byte receive and transmit
+ *       first-in/first-out (FIFO) buffer, however some specialty
+ *       controller cards allowed up to 64-bytes and later 256-bytes.
+ */
 
 class MyFifo {
 public:
-	MyFifo(Bitu maxsize_) {
-		maxsize=size=maxsize_;
-		pos=used=0;
-		data=new Bit8u[size];
-	}
-	~MyFifo() {
-		delete[] data;
-	}
-	INLINE Bitu getFree(void) {
-		return size-used;
-	}
-	bool isEmpty() {
-		return used==0;
-	}
-	bool isFull() {
-		return (size-used)==0;
-	}
+	MyFifo(const MyFifo &) = delete;            // prevent copying
+	MyFifo &operator=(const MyFifo &) = delete; // prevent assignment
 
-	INLINE Bitu getUsage(void) {
-		return used;
+	MyFifo(size_t n) : data(n, 0), maxsize(n), size(n)
+	{
+		assert(n <= SERIAL_MAX_FIFO_SIZE);
 	}
-	void setSize(Bitu newsize)
+	size_t getFree() { return size - used; }
+	bool isEmpty() { return used == 0; }
+	bool isFull() { return (size - used) == 0; }
+	size_t getUsage() { return used; }
+	void setSize(size_t n)
 	{
-		size=newsize;
-		pos=used=0;
+		assert(n <= SERIAL_MAX_FIFO_SIZE);
+		data.resize(n);
+		size = n;
+		maxsize = n;
+		pos = used = 0;
 	}
-	void clear(void) {
-		pos=used=0;
-		data[0]=0;
+	void clear() {
+		pos = used = 0;
+		if (data.size() > 0)
+			std::fill(data.begin(), data.end(), 0);
 	}
 
-	bool addb(Bit8u _val) {
-		Bitu where=pos+used;
-		if (where>=size) where-=size;
-		if(used>=size) {
+	bool addb(uint8_t val)
+	{
+		size_t where = pos + used;
+		if (where >= size)
+			where -= size;
+		if (used >= size) {
 			// overwrite last byte
-			if(where==0) where=size-1;
+			if (where == 0)
+				where = size - 1;
 			else where--;
-			data[where]=_val;
+			assert(where < data.size());
+			data[where] = val;
 			return false;
 		}
-		data[where]=_val;
+		assert(where < data.size());
+		data[where] = val;
 		used++;
 		return true;
```

Comment:
All logic is the same (just reformatted).
Only added two `assert`s to make sure we don't read beyond the array length.